### ISSUE TITLE
[Feature] Create an "OverrideApplication" CRD to manipulate an existing Application in the cluster

### ISSUE THREAD

**Issue by YourTechBud** [kubevela/kubevela#3871]
**Is your feature request related to a problem? Please describe.**
Many times you would want to modify an application without having to change the backing "Application" object. For example, to create a preview environment for an application , i may want to deploy an additional deployment and modify the ingress defined by that application. I can do that by modifying the backing "Application" object itself, but then it may cause a lot of problems during automation.

**Describe the solution you'd like**
I suggest we create a "ApplicationOverrides" crd which the kubevela operator can merge with the original Application object before applying. This way I can create a layers application description with the main "Application" object serving as the base while each "ApplicationOverrides" acts as an overlay.

**Describe alternatives you've considered**
I tried using helm for this but it turns out that helm makes it very difficult to merge arrays (components and traits are arrays).

My current solution still uses helm where i have created an map called "overrides" which archives the same result. Each overrides goes in a seperate values.yaml file. Its a very hacky solution and i was hoping kubevela will be able to solve this natively.

**Additional context**
Need this feature to power the "preview environments" feature where we create a deployment with an experimental feature and allow traffic containing a specific header / url to access it

**Comment by wonderflow** [comment]
Do you mean the `OverrideApplication` won't affect the current application and it will run as another application in another K8s cluster or namespace?

**Comment by YourTechBud** [comment]
No no. "ApplicationOverrides" will be merged with the existing Application which it is overriding. Whether it creates a new deployment or not is completely upto its contents.

I imagine the "ApplicationOverrides" object will be used to achieve one or more of the following:
- Add new components to an existing Application
- Add new traits to an existing component of the Application
- Modify the parameters of an existing trait / component
- remove components or traits from an existing Application.(i don't this this would be used at all)

In Kubevela terms, I'd think of this "ApplicationOverrides" object as a patcher which tries to patch an existing Application.

**Comment by wonderflow** [comment]
If the main goal of `ApplicationOverrides` is to modify the current application and takes affect to the underlying resources, it can cause lot's of confusion and make the ecosystem hard to integrate especially for the observability.

**Comment by YourTechBud** [comment]
Yes. That makes sense. But we can have a command to print the current active version of the Application with all the Overrides applied?

Thinking about it, Kustomise can solve this problem as well since it has the concept of a base and layers too. Adding the "ApplicationOverrides" featute will add Kustomise like functionality to Kubevela.

I was hoping i didn't have to use another tool and get those features directly

**Comment by zhiyuone** [comment]
This looks like a multi-environment configuration feature, we have [Override](https://kubevela.net/zh/docs/next/end-user/policies/references#override) policy, can it meet this need?

**Comment by YourTechBud** [comment]
> This looks like a multi-environment configuration feature, we have Override policy, can it meet this need?

I think this will work. Thanks for sharing @zhiyuone!! Can we use this to even add a component or can i just patch an existing one?

**Comment by Somefive** [comment]
> > This looks like a multi-environment configuration feature, we have Override policy, can it meet this need?
> 
> I think this will work. Thanks for sharing @zhiyuone!! Can we use this to even add a component or can i just patch an existing one?

Adding new component is possible if the component to override in the policy has a name different than all the existing components.

**Comment by YourTechBud** [comment]
> This looks like a multi-environment configuration feature, we have [Override](https://kubevela.net/zh/docs/next/end-user/policies/references#override) policy, can it meet this need?

Hey @Somefive & @zhiyuone. Unfortunately the override policy does not solve my problem. Here is my scenario in more detail:

- All kubevela `Applications` are being deployed via kubectl. Each  Kubevela `Application` creates a `Deployment`, `Service` and `Istio VirtualService`
- My developer wants to debug an application directly in the development environment. We use [telepresence.io](https://www.telepresence.io/) to achieve this. However, telepresence intercepts the entire deployment which affects other microservices in the cluster.
- So instead of using telepresence directly, my developer runs a script which creates a new `dummy` deployment, intercepts that using telepresence and then modifies the istio virtual service to direct traffic containing a particular header to the intercepted deployment.
- The problem is that modifying the virtual service makes kubevela angry since the virtual service was originally deployed by an Kubevela application. So in Kubevela I have to use the apply once policy which I'm not particularly happy with.

An external `ApplicationOverride` resource will allow me to modify the virtual service (and create that temporary deployment) in a way kubevela is aware of. Moreover it will not disturb my existing CD pipeline in any way.

**Comment by briankane** [comment]
This issue has been inactive for over 2 years and is scheduled for closure as part of our backlog maintenance.

  **If this issue is still relevant:**
  - Please comment below to keep it open
  - Provide updated context if the original description is outdated

  **Timeline:** If there's no activity by **Friday, October 24th**, this issue will be automatically closed.

  **Note:** Issues can always be reopened if they become relevant again. We encourage the community to reopen and reprioritize any incorrectly 
  closed issues.

**Comment by briankane** [comment]
This issue has been closed following the stale issue notice. 

If this issue is still relevant and should remain open, please:
1. Reopen the issue
2. Remove the 'stale' label
3. Provide an update on the current status or importance

Thank you for your understanding and contribution to KubeVela!
